README.md
# R-package 'multiview'

Methods for high-dimensional multi-view learning. The multiview package currently provides functions to fit stacked penalized logistic regression (StaPLR) models, which are a special case of multi-view stacking (MVS).

For more information about the StaPLR and MVS methods, see https://doi.org/10.1016/j.inffus.2020.03.007.

## Installation

Using the 'devtools' package:

~~~ r
devtools::install_gitlab("wsvanloon/multiview")
~~~

## Using 'multiview'

The two main functions are:

- StaPLR() - fits penalized and stacked penalized logistic regression models models up to two levels. Can be used to generate cross-validated predictions.
- MVS() - fits StaPLR models with >= 2 levels.

Objects returned by either function have associated coef() and predict() methods. 

CHANGELOG.txt
February 23, 2021 - Version 0.3.1
- MVS() now correctly applies nonnegativity constraints at the intermediate level(s) if specified.
- The MVS() example has been modified slightly.

February 4, 2021 - Version 0.3.0
- multiview is now licenced under GNU GPLv2
- Added a new function MVS() which provides a more modular approach to multi-view stacking.
- MVS() allows the application of StaPLR with more than 2 levels.
- MVS() outputs objects of S3 class MVS which have associated coef() and predict() methods.
- Added an option to StaPLR to skip training the meta-learner. The meta-learner is automatically skipped when only one view is provided.
- Added an option to StaPLR to skip the generation of cross-validated predictions.
- StaPLR output no longer contains the original training data.
- The default number of folds for StaPLR is now 10 (was 5)
- The default value of lambda.ratio is now 1e-04 (was 0.01).
- By default, StaPLR no longer checks the version of glmnet, nor does it make suggestions on which version to use. If desired, the version check can still be enabled by using StaPLR(..., skip.version=TRUE).
- The R project file has been removed and is no longer tracked in the repository.



R/MVS.R
# This file is part of multiview: Methods for High-Dimensional Multi-View Learning
# Copyright (C) 2018-2021  Wouter van Loon
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

#' Multi-View Stacking
#'
#' Fit a multi-view stacking model with two or more levels.
#' @param x input matrix of dimension nobs x nvars.
#' @param y outcome vector of length nobs.
#' @param views a matrix of dimension nvars x (levels - 1), where each entry is an integer describing to which view each feature corresponds.
#' @param type the type of MVS model to be fitted. Currently only type "StaPLR" is supported.
#' @param levels an integer >= 2, specifying the number of levels in the MVS procedure.
#' @param alphas a vector specifying the value of the alpha parameter to use at each level.
#' @param nnc a binary vector specifying whether to apply nonnegativity constraints or not (1/0) at each level.
#' @param parallel whether to use foreach to fit the learners and obtain the cross-validated predictions at each level in parallel. Executes sequentially unless a parallel back-end is registered beforehand.
#' @param seeds (optional) a vector specifying the seed to use at each level.
#' @param progress whether to show a progress bar (only supported when parallel = FALSE).
#' @param ... additional arguments to pass to the learning algorithm. See e.g. ?StaPLR. Note that these arguments are passed to the the learner at every level of the MVS procedure.
#' @return An object of S3 class "MVS".
#' @keywords TBA
#' @import foreach
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' X <- matrix(rnorm(8500), nrow=n, ncol=85)
#' top_level <- c(rep(1,45), rep(2,20), rep(3,20))
#' bottom_level <- c(rep(1:3, each=15), rep(4:5, each=10), rep(6:9, each=5))
#' views <- cbind(bottom_level, top_level)
#' beta <- c(rep(10, 55), rep(0, 30)) * ((rbinom(85, 1, 0.5)*2)-1)
#' eta <- X %*% beta
#' p <- 1 /(1 + exp(-eta))
#' y <- rbinom(n, 1, p)
#'
#' fit <- MVS(x=X, y=y, views=views, type="StaPLR", levels=3, alphas=c(0,1,1), nnc=c(0,1,1))
#' coefficients <- coef(fit)
#'
#' new_X <- matrix(rnorm(2*85), nrow=2)
#' predict(fit, new_X)

MVS <- function(x, y, views, type="StaPLR", levels=2, alphas=c(0,1), nnc=c(0,1), parallel=FALSE, seeds=NULL, progress=TRUE, ...){

  pred_functions <- vector("list", length=ncol(views)+1)
  ll <- c(-Inf, 0)[nnc + 1]

  if(progress){
    message("Level 1 \n")
  }

  pred_functions[[1]] <- learn(X=x, y=y, views=views[,1], type=type, alpha1 = alphas[1], ll1=ll[1],
                               seed=seeds[1], progress=progress, parallel=parallel, ...)

  if(levels > 2){
    for(i in 2:ncol(views)){
      if(progress){
        message(paste("Level", i, "\n"))
      }
      pred_functions[[i]] <- learn(pred_functions[[i-1]]$CVs, y,
                                   views=condense(views, level=i), type=type,
                                   alpha1 = alphas[i], ll1=ll[i], seed=seeds[i],
                                   progress=progress, parallel=parallel, ...)
    }
  }

  if(progress){
    message(paste("Level", ncol(views)+1, "\n"))
  }

  pred_functions[[ncol(views)+1]] <- learn(pred_functions[[ncol(views)]]$CVs, y,
                                           views=rep(1,ncol(pred_functions[[ncol(views)]]$CVs)),
                                           type=type, alpha1 = alphas[ncol(views)+1], ll1=ll[ncol(views)+1],
                                           generate.CVs=FALSE, seed=seeds[ncol(views)+1],
                                           progress=progress, parallel=parallel, ...)

  for(i in 1:length(pred_functions)){
    pred_functions[[i]]$meta <- NULL
    names(pred_functions[[i]])[1] <- "models"
  }

  names(pred_functions) <- paste("Level", 1:(ncol(views)+1))
  attr(pred_functions, "type") <- type
  class(pred_functions) <- "MVS"

  return(pred_functions)

}


#' Make predictions from an "MVS" object.
#'
#' Make predictions from a "MVS" object.
#' @param object An object of class "MVS".
#' @param newx Matrix of new values for x at which predictions are to be made. Must be a matrix.
#' @param predtype The type of prediction returned by the meta-learner. Supported are types "response", "class" and "link".
#' @param cvlambda Values of the penalty parameters at which predictions are to be made. Defaults to the values giving minimum cross-validation error.
#' @return A matrix of predictions.
#' @keywords TBA
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' X <- matrix(rnorm(8500), nrow=n, ncol=85)
#' top_level <- c(rep(1,45), rep(2,20), rep(3,20))
#' bottom_level <- c(rep(1:3, each=15), rep(4:5, each=10), rep(6:9, each=5))
#' views <- cbind(bottom_level, top_level)
#' beta <- c(rep(10, 55), rep(0, 30)) * ((rbinom(85, 1, 0.5)*2)-1)
#' eta <- X %*% beta
#' p <- 1 /(1 + exp(-eta))
#' y <- rbinom(n, 1, p)
#'
#' fit <- MVS(x=X, y=y, views=views, type="StaPLR", levels=3, alphas=c(0,1,1), nnc=c(0,1,1))
#' coefficients <- coef(fit)
#'
#' new_X <- matrix(rnorm(2*85), nrow=2)
#' predict(fit, new_X)

predict.MVS <- function(object, newx, predtype = "response", cvlambda = "lambda.min"){

  x <- newx

  for(i in 1:length(object)){

    Z <- matrix(NA, nrow=nrow(newx), ncol=length(object[[i]]$models))

    if(i < length(object)){
      pt <- object[[i]]$metadat
    }
    else pt <- predtype

    for(j in 1:ncol(Z)){
      Z[,j] <- predict(object[[i]]$models[[j]], x[, object[[i]]$view == j, drop=FALSE],
                       s = cvlambda, type = pt)
    }
    x <- Z
  }

  return(x)

}


#' Extract coefficients from an "MVS" object.
#'
#' Extract coefficients at each level from an "MVS" object at the CV-optimal values of the penalty parameters.
#' @param object An object of class "MVS".
#' @param cvlambda By default, the coefficients are extracted at the CV-optimal values of the penalty parameters. Choosing "lambda.1se" will extract them at the largest values within one standard error of the minima.
#' @return An object of S3 class "MVScoef".
#' @keywords TBA
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' X <- matrix(rnorm(8500), nrow=n, ncol=85)
#' top_level <- c(rep(1,45), rep(2,20), rep(3,20))
#' bottom_level <- c(rep(1:3, each=15), rep(4:5, each=10), rep(6:9, each=5))
#' views <- cbind(bottom_level, top_level)
#' beta <- c(rep(10, 55), rep(0, 30)) * ((rbinom(85, 1, 0.5)*2)-1)
#' eta <- X %*% beta
#' p <- 1 /(1 + exp(-eta))
#' y <- rbinom(n, 1, p)
#'
#' fit <- MVS(x=X, y=y, views=views, type="StaPLR", levels=3, alphas=c(0,1,1), nnc=c(0,1,1))
#' coefficients <- coef(fit)
#'
#' new_X <- matrix(rnorm(2*85), nrow=2)
#' predict(fit, new_X)

coef.MVS <- function(object, cvlambda = "lambda.min"){

  out <- vector("list", length(object))

  for(i in 1:length(object)){
    out[[i]] <- vector("list", length(object[[i]]$models))
  }

  for(i in 1:length(object)){
    for(j in 1:length(object[[i]]$models)){
      out[[i]][[j]] <- coef(object[[i]]$models[[j]], s=cvlambda)
    }
  }
  names(out) <- paste("Level", 1:length(object))
  attr(out, "type") <- attr(object, "type")
  class(out) <- "MVScoef"

  return(out)
}


R/learn.R
# This file is part of multiview: Methods for High-Dimensional Multi-View Learning
# Copyright (C) 2018-2021  Wouter van Loon
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# Learn is a generic function to train a learner on multi-view input data,
# and produce learned functions and (optionally) the matrix of cross-validated
# predictions. It uses switch() to apply the correct learner, depending on
# argument 'type'.

learn <- function(X, y, views, type, generate.CVs=TRUE, ...){
  switch(type,
         StaPLR = StaPLR(X, y, view=views, skip.meta = TRUE, skip.cv = !generate.CVs, ...))
}


R/StaPLR3.R
# This is a quick-and-dirty script to apply StaPLR with 3 levels. Its outcome can potentially be used for automated testing of MVS later on.

StaPLR3 <- function(X, y , views, alphas, nfolds=10, seed=NULL){

  base_learners <- StaPLR(X, y, view=views[,1], alpha1 = alphas[1], alpha2 = alphas[2], skip.meta=TRUE, ll1 = -Inf, ul1 = Inf, seed=seed)

  trans_learners <- StaPLR(base_learners$CVs, y, view=condense(views, level=2), alpha1 = alphas[2], alpha2 = alphas[3], ll1 = 0, ul1 = Inf,
                           ll2 = 0, ul2 = Inf, seed=seed)

  out <- list(base=base_learners$base,
              trans=trans_learners$base,
              meta=trans_learners$meta,
              Z1=base_learners$CVs,
              Z2=trans_learners$CVs,
              X=base_learners$x,
              y=base_learners$y,
              views1=base_learners$view,
              views2=trans_learners$view)

  class(out) <- "StaPLR3"

  return(out)
}


condense <- function(views, level){

  tab <- xtabs(~views[, level - 1] + views[, level])
  condensed_views <- rep(NA, length(unique(views[, level])))
  for(i in 1:nrow(tab)){
    condensed_views [i] <- which(tab[i,] != 0)
  }

  return(condensed_views)
}


predict.StaPLR3 <- function(object, newx, newcf = NULL, predtype = "response", cvlambda = "lambda.min"){

  V1 <- length(unique(object$views1))
  n <- nrow(newx)
  metadat <- "response" # THIS SHOULD INHERIT FROM STAPLR3 OBJECT INSTEAD!
  Z1 <- matrix(NA, n, V1)
  for (v in 1:V1){
    Z1[,v] <- predict(object$base[[v]], newx[, object$views1 == v, drop=FALSE], s = cvlambda, type = metadat)
  }
  # if(!is.null(newcf)){
  #   Z <- cbind(newcf, Z)
  # }
  colnames(Z1) <- colnames(object$CVs)

  V2 <- length(unique(object$views2))
  Z2 <- matrix(NA, n, V2)
  for (v in 1:V2){
    Z2[,v] <- predict(object$trans[[v]], Z1[, object$views2 == v, drop=FALSE], s = cvlambda, type = metadat)
  }

  out <- predict(object$meta, Z2, s = cvlambda, type = predtype)
  return(out)

}

# TEST DATA
# n <- 1000
# set.seed(111220)
# test_X<- matrix(rnorm(8500), nrow=n, ncol=85)
# top_level <- c(rep(1,45), rep(2,20), rep(3,20))
# bottom_level <- c(rep(1:3, each=15), rep(4:5, each=10), rep(6:9, each=5))
# views <- cbind(bottom_level, top_level)
#
# beta <- c(rep(10, 55), rep(0, 30))
# eta <- test_X %*% beta
# p <- 1 /(1 + exp(-eta))
# test_y <- rbinom(n, 1, p)


R/condense.R
# This file is part of multiview: Methods for High-Dimensional Multi-View Learning
# Copyright (C) 2018-2021  Wouter van Loon
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

condense <- function(views, level){

  tab <- xtabs(~views[, level - 1] + views[, level])
  condensed_views <- rep(NA, length(unique(views[, level])))
  for(i in 1:nrow(tab)){
    condensed_views [i] <- which(tab[i,] != 0)
  }

  return(condensed_views)
}


R/StaPLR.R
# This file is part of multiview: Methods for High-Dimensional Multi-View Learning
# Copyright (C) 2018-2021  Wouter van Loon
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

#' Stacked Penalized Logistic Regression
#'
#' Fit a two-level stacked penalized logistic regression model with a single base-learner and a single meta-learner.
#' @param x input matrix of dimension nobs x nvars
#' @param y outcome vector of length nobs
#' @param view a vector of length nvars, where each entry is an integer describing to which view each feature corresponds.
#' @param view.names (optional) a character vector of length nviews specifying a name for each view.
#' @param correct.for (optional) a matrix with nrow = nobs, where each column is a feature which should be included directly into the meta.learner. By default these features are not penalized (see penalty.weights) and appear at the top of the coefficient list.
#' @param alpha1 (base) alpha parameter for glmnet: lasso(1) / ridge(0)
#' @param alpha2 (meta) alpha parameter for glmnet: lasso(1) / ridge(0)
#' @param nfolds number of folds to use for all cross-validation.
#' @param seed (optional) numeric value specifying the seed. Setting the seed this way ensures the results are reproducable even when the computations are performed in parallel.
#' @param std.base should features be standardized at the base level?
#' @param std.meta should cross-validated predictions be standardized at the meta level?
#' @param ll1 lower limit(s) for each coefficient at the base-level. Defaults to -Inf.
#' @param ul1 upper limit(s) for each coefficient at the base-level. Defaults to Inf.
#' @param ll2 lower limit(s) for each coefficient at the meta-level. Defaults to 0 (non-negativity constraints). Does not apply to correct.for features.
#' @param ul2 upper limit(s) for each coefficient at the meta-level. Defaults to Inf. Does not apply to correct.for features.
#' @param cvloss loss to use for cross-validation.
#' @param metadat which attribute of the base learners should be used as input for the meta learner? Allowed values are "response", "link", and "class".
#' @param cvlambda value of lambda at which cross-validated predictions are made. Defaults to the value giving minimum internal cross-validation error.
#' @param cvparallel whether to use 'foreach' to fit each CV fold (DO NOT USE, USE OPTION parallel INSTEAD).
#' @param lambda.ratio the ratio between the largest and smallest lambda value.
#' @param penalty.weights (optional) a vector of length nviews, containing different penalty factors for the meta-learner. Defaults to rep(1,nviews). The penalty factor is set to 0 for correct.for features.
#' @param parallel whether to use foreach to fit the base-learners and obtain the cross-validated predictions in parallel. Executes sequentially unless a parallel backend is registered beforehand.
#' @param skip.fdev whether to skip checking if the fdev parameter is set to zero.
#' @param skip.version whether to skip checking the version of the glmnet package.
#' @param skip.meta whether to skip training the metalearner.
#' @param skip.cv whether to skip generating the cross-validated predictions.
#' @param progress whether to show a progress bar (only supported when parallel = FALSE).
#' @return An object with S3 class "StaPLR".
#' @keywords TBA
#' @import foreach
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' cors <- seq(0.1,0.7,0.1)
#' X <- matrix(NA, nrow=n, ncol=length(cors)+1)
#' X[,1] <- rnorm(n)
#'
#' for(i in 1:length(cors)){
#'   X[,i+1] <- X[,1]*cors[i] + rnorm(n, 0, sqrt(1-cors[i]^2))
#' }
#'
#' beta <- c(1,0,0,0,0,0,0,0)
#' eta <- X %*% beta
#' p <- exp(eta)/(1+exp(eta))
#' y <- rbinom(n, 1, p)
#' view_index <- rep(1:(ncol(X)/2), each=2)
#'
#' fit <- StaPLR(X, y, view_index)
#' coef(fit)$meta
#'
#' new_X <- matrix(rnorm(16), nrow=2)
#' predict(fit, new_X)

StaPLR <- function(x, y, view, view.names = NULL, correct.for = NULL, alpha1 = 0, alpha2 = 1, nfolds = 10, seed = NULL,
                      std.base = FALSE, std.meta = FALSE, ll1 = -Inf, ul1 = Inf,
                      ll2 = 0, ul2 = Inf, cvloss = "deviance", metadat = "response", cvlambda = "lambda.min",
                      cvparallel = FALSE, lambda.ratio = 1e-4, penalty.weights = NULL, parallel = FALSE, skip.fdev = FALSE, skip.version = TRUE, skip.meta = FALSE, skip.cv = FALSE, progress = TRUE){

  # Check if glmnet.control parameter fdev is set to zero.
  if(skip.fdev == FALSE){
    if(glmnet.control()$fdev != 0){
      glmnet.control(fdev=0)
      message("Minimum fractional change in deviance for stopping path set to zero for the duration of this session. \n To reset to default use: glmnet.control(factory=TRUE) (NOT RECOMMENDED). \n To skip this check use: StaPLR(..., skip.fdev=TRUE) (NOT RECOMMENDED). \n")
    }
  }

  # Check current version of glmnet
  if(skip.version == FALSE){
    versionMessage <- paste0("Found glmnet version ", packageVersion("glmnet"), ".\n To skip this check use: StaPLR(..., skip.version=TRUE). \n")
    message(versionMessage)
  }

  # object initialization
  V <- length(unique(view))
  n <- length(y)

  if(V==1 && !skip.meta){
    warning("Only 1 view was provided. Training the meta-learner will be skipped!")
    skip.meta <- TRUE
  }

  # SEQUENTIAL PROCESSING
  if(parallel == FALSE){

    if(!is.null(seed)){
      set.seed(seed)
      folds <- kFolds(y, nfolds)
      base.seeds <- sample(.Machine$integer.max/2, size = V)
      z.seeds <- matrix(sample(.Machine$integer.max/2, size = V*nfolds), nrow=nfolds, ncol=V)
      meta.seed <- sample(.Machine$integer.max/2, size=1)
    }
    else folds <- kFolds(y, nfolds)

    if(progress == TRUE){
      message("Training learner on each view...")
      pb <- txtProgressBar(min=0, max=V, style=3)
    }
    cv.base <- foreach(v=(1:V)) %do% {
      if(progress == TRUE){
        setTxtProgressBar(pb, v)
      }
      if(!is.null(seed)){
        set.seed(base.seeds[v])
      }
      glmnet::cv.glmnet(x[, view == v], y, family = "binomial", nfolds = nfolds,
                        type.measure = cvloss, alpha = alpha1,
                        standardize = std.base, lower.limits = ll1,
                        upper.limits = ul1, parallel = cvparallel, lambda.min.ratio = lambda.ratio)
    }

    if(!skip.cv){
      if(progress == TRUE){
        message("\n Calculating cross-validated predictions...")
        pb <- txtProgressBar(min=0, max=V*nfolds, style=3)
      }

      Z <- foreach(v=(1:V), .combine=cbind) %:%
        foreach(k=(1:nfolds), .combine="+") %do% {
          if(progress == TRUE){
            setTxtProgressBar(pb, getTxtProgressBar(pb)+1)
          }
          if(!is.null(seed)){
            set.seed(z.seeds[k, v])
          }
          cvf <- glmnet::cv.glmnet(x[folds != k, view == v], y[folds != k], family = "binomial", nfolds = nfolds,
                                   type.measure = cvloss, alpha = alpha1,
                                   standardize = std.base, lower.limits = ll1,
                                   upper.limits = ul1, parallel = cvparallel, lambda.min.ratio = lambda.ratio)
          newy <- rep(0, length(y))
          newy[folds == k] <- predict(cvf, newx = x[folds == k, view == v], s = cvlambda, type = metadat)
          return(newy)
        }
      dimnames(Z) <- NULL
      if(!is.null(view.names)){
        colnames(Z) <- view.names
      }
    }
    else{
      Z <- NULL
      skip.meta <- TRUE
    }

    if(progress == TRUE && skip.meta == FALSE){
      message("\n Training meta learner...")
    }
    if(!is.null(seed)){
      set.seed(meta.seed)
    }
    if(skip.meta == TRUE){
      cv.meta <- NULL
    }
    else if(is.null(correct.for) && is.null(penalty.weights)){
      cv.meta <- glmnet::cv.glmnet(Z, y, family= "binomial", nfolds = nfolds, type.measure = cvloss, alpha = alpha2,
                                   standardize = std.meta, lower.limits = ll2,
                                   upper.limits = ul2, parallel = cvparallel, lambda.min.ratio=lambda.ratio)
    }
    else if(is.null(correct.for) && !is.null(penalty.weights)){
      cv.meta <- glmnet::cv.glmnet(Z, y, family= "binomial", nfolds = nfolds, type.measure = cvloss, alpha = alpha2,
                                   standardize = std.meta, lower.limits = ll2,
                                   upper.limits = ul2, parallel = cvparallel, lambda.min.ratio=lambda.ratio, penalty.factor=penalty.weights)
    }
    else{
      if(is.null(penalty.weights)){
        penalty.weights <- c(rep(0, ncol(correct.for)), rep(1, ncol(Z)))
      }
      else{
        penalty.weights <- c(rep(0, ncol(correct.for)), penalty.weights)
      }
      ll2 <- c(rep(-Inf, ncol(correct.for)), rep(ll2, ncol(Z)))
      ul2 <- c(rep(Inf, ncol(correct.for)), rep(ul2, ncol(Z)))
      Z <- cbind(correct.for, Z)
      cv.meta <- glmnet::cv.glmnet(Z, y, family= "binomial", nfolds = nfolds, type.measure = cvloss, alpha = alpha2,
                                   standardize = std.meta, lower.limits = ll2,
                                   upper.limits = ul2, parallel = cvparallel, lambda.min.ratio=lambda.ratio, penalty.factor=penalty.weights)
    }

  }

  # PARALLEL PROCESSING
  if(parallel == TRUE){

    if(!is.null(seed)){
      set.seed(seed)
      folds <- kFolds(y, nfolds)
      base.seeds <- sample(.Machine$integer.max/2, size = V)
      z.seeds <- matrix(sample(.Machine$integer.max/2, size = V*nfolds), nrow=nfolds, ncol=V)
      meta.seed <- sample(.Machine$integer.max/2, size=1)
    }
    else folds <- kFolds(y, nfolds)

    cv.base <- foreach(v=(1:V)) %dopar% {
      if(!is.null(seed)){
        set.seed(base.seeds[v])
      }
      glmnet::cv.glmnet(x[, view == v], y, family = "binomial", nfolds = nfolds,
                        type.measure = cvloss, alpha = alpha1,
                        standardize = std.base, lower.limits = ll1,
                        upper.limits = ul1, parallel = cvparallel, lambda.min.ratio = lambda.ratio)
    }

    if(!skip.cv){
      Z <- foreach(v=(1:V), .combine=cbind) %:%
        foreach(k=(1:nfolds), .combine="+") %dopar% {
          if(!is.null(seed)){
            set.seed(z.seeds[k, v])
          }
          cvf <- glmnet::cv.glmnet(x[folds != k, view == v], y[folds != k], family = "binomial", nfolds = nfolds,
                                   type.measure = cvloss, alpha = alpha1,
                                   standardize = std.base, lower.limits = ll1,
                                   upper.limits = ul1, parallel = cvparallel, lambda.min.ratio = lambda.ratio)
          newy <- rep(0, length(y))
          newy[folds == k] <- predict(cvf, newx = x[folds == k, view == v], s = cvlambda, type = metadat)
          return(newy)
        }
      dimnames(Z) <- NULL
      if(!is.null(view.names)){
        colnames(Z) <- view.names
      }
    }
    else{
      Z <- NULL
      skip.meta <- TRUE
    }

    if(!is.null(seed)){
      set.seed(meta.seed)
    }
    if(skip.meta == TRUE){
      cv.meta <- NULL
    }
    else if(is.null(correct.for) && is.null(penalty.weights)){
      cv.meta <- glmnet::cv.glmnet(Z, y, family= "binomial", nfolds = nfolds, type.measure = cvloss, alpha = alpha2,
                                   standardize = std.meta, lower.limits = ll2,
                                   upper.limits = ul2, parallel = cvparallel, lambda.min.ratio=lambda.ratio)
    }
    else if(is.null(correct.for) && !is.null(penalty.weights)){
      cv.meta <- glmnet::cv.glmnet(Z, y, family= "binomial", nfolds = nfolds, type.measure = cvloss, alpha = alpha2,
                                   standardize = std.meta, lower.limits = ll2,
                                   upper.limits = ul2, parallel = cvparallel, lambda.min.ratio=lambda.ratio, penalty.factor=penalty.weights)
    }
    else{
      if(is.null(penalty.weights)){
        penalty.weights <- c(rep(0, ncol(correct.for)), rep(1, ncol(Z)))
      }
      else{
        penalty.weights <- c(rep(0, ncol(correct.for)), penalty.weights)
      }
      ll2 <- c(rep(-Inf, ncol(correct.for)), rep(ll2, ncol(Z)))
      ul2 <- c(rep(Inf, ncol(correct.for)), rep(ul2, ncol(Z)))
      Z <- cbind(correct.for, Z)
      cv.meta <- glmnet::cv.glmnet(Z, y, family= "binomial", nfolds = nfolds, type.measure = cvloss, alpha = alpha2,
                                   standardize = std.meta, lower.limits = ll2,
                                   upper.limits = ul2, parallel = cvparallel, lambda.min.ratio=lambda.ratio, penalty.factor=penalty.weights)
    }
  }


  # create output list
  out <- list(
    "base" = cv.base,
    "meta" = cv.meta,
    "CVs" = Z,
    #"x" = x,
    #"y" = y,
    "view" = view,
    "metadat" = metadat
  )

  class(out) <- "StaPLR"

  # return output
  message("DONE")
  return(out)
}



#' Make predictions from a "StaPLR" object.
#'
#' Make predictions from a "StaPLR" object.
#' @param object Fitted "StaPLR" model object.
#' @param newx Matrix of new values for x at which predictions are to be made. Must be a matrix.
#' @param newcf Matrix of new values of correction features, if correct.for was specified during model fitting.
#' @param predtype The type of prediction returned by the meta-learner.
#' @param cvlambda Values of the penalty parameters at which predictions are to be made. Defaults to the values giving minimum cross-validation error.
#' @return A matrix of predictions.
#' @keywords TBA
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' cors <- seq(0.1,0.7,0.1)
#' X <- matrix(NA, nrow=n, ncol=length(cors)+1)
#' X[,1] <- rnorm(n)
#'
#' for(i in 1:length(cors)){
#'   X[,i+1] <- X[,1]*cors[i] + rnorm(n, 0, sqrt(1-cors[i]^2))
#' }
#'
#' beta <- c(1,0,0,0,0,0,0,0)
#' eta <- X %*% beta
#' p <- exp(eta)/(1+exp(eta))
#' y <- rbinom(n, 1, p)
#' view_index <- rep(1:(ncol(X)/2), each=2)
#'
#' fit <- StaPLR(X, y, view_index)
#' coef(fit)$meta
#'
#' new_X <- matrix(rnorm(16), nrow=2)
#' predict(fit, new_X)

predict.StaPLR <- function(object, newx, newcf = NULL, predtype = "response", cvlambda = "lambda.min"){

  V <- length(unique(object$view))
  n <- nrow(newx)
  metadat <- object$metadat
  Z <- matrix(NA, n, V)
  for (v in 1:V){
    Z[,v] <- predict(object$base[[v]], newx[, object$view == v, drop=FALSE], s = cvlambda, type = metadat)
  }
  if(!is.null(newcf)){
    Z <- cbind(newcf, Z)
  }
  colnames(Z) <- colnames(object$CVs)
  out <- predict(object$meta, Z, s = cvlambda, type = predtype)
  return(out)
}



#' Extract coefficients from a "StaPLR" object.
#'
#' Extract base- and meta-level coefficients from a "StaPLR" object at the CV-optimal values of the penalty parameters.
#' @param object Fitted "StaPLR" model object.
#' @param cvlambda By default, the coefficients are extracted at the CV-optimal values of the penalty parameters. Choosing "lambda.1se" will extract them at the largest values within one standard error of the minima.
#' @return An object with S3 class "StaPLRcoef".
#' @keywords TBA
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' cors <- seq(0.1,0.7,0.1)
#' X <- matrix(NA, nrow=n, ncol=length(cors)+1)
#' X[,1] <- rnorm(n)
#'
#' for(i in 1:length(cors)){
#'   X[,i+1] <- X[,1]*cors[i] + rnorm(n, 0, sqrt(1-cors[i]^2))
#' }
#'
#' beta <- c(1,0,0,0,0,0,0,0)
#' eta <- X %*% beta
#' p <- exp(eta)/(1+exp(eta))
#' y <- rbinom(n, 1, p)
#' view_index <- rep(1:(ncol(X)/2), each=2)
#'
#' fit <- StaPLR(X, y, view_index)
#' coef(fit)$meta
#'
#' new_X <- matrix(rnorm(16), nrow=2)
#' predict(fit, new_X)

coef.StaPLR <- function(object, cvlambda = "lambda.min"){

  out <- list(
    "base" = lapply(object$base, function(x) coef(x, s=cvlambda)),
    "meta" = coef(object$meta, s=cvlambda),
    "metadat" = object$metadat
  )

  class(out) <- "StaPLRcoef"

  return(out)
}


#' Make predictions from a "StaPLRcoef" object.
#'
#' Predict using a "StaPLRcoef" object. A "StaPLRcoef" object can be considerably smaller than a full "StaPLR" object for large data sets.
#' @param object Extracted StaPLR coefficients as a "StaPLRcoef" object.
#' @param newx Matrix of new values for x at which predictions are to be made. Must be a matrix.
#' @param view a vector of length nvars, where each entry is an integer describing to which view each feature corresponds.
#' @param newcf Matrix of new values of correction features, if correct.for was specified during model fitting.
#' @param predtype The type of prediction returned by the meta-learner. Allowed values are "response", "link", and "class".
#' @return A matrix of predictions.
#' @keywords TBA
#' @export
#' @author Wouter van Loon <w.s.van.loon@fsw.leidenuniv.nl>
#' @examples
#' set.seed(012)
#' n <- 1000
#' cors <- seq(0.1,0.7,0.1)
#' X <- matrix(NA, nrow=n, ncol=length(cors)+1)
#' X[,1] <- rnorm(n)
#'
#' for(i in 1:length(cors)){
#'   X[,i+1] <- X[,1]*cors[i] + rnorm(n, 0, sqrt(1-cors[i]^2))
#' }
#'
#' beta <- c(1,0,0,0,0,0,0,0)
#' eta <- X %*% beta
#' p <- exp(eta)/(1+exp(eta))
#' y <- rbinom(n, 1, p)
#' view_index <- rep(1:(ncol(X)/2), each=2)
#'
#' fit <- StaPLR(X, y, view_index)
#' coefficients <- coef(fit)
#'
#' new_X <- matrix(rnorm(16), nrow=2)
#' predict(coefficients, new_X, view_index)

predict.StaPLRcoef <- function(object, newx, view, newcf = NULL, predtype = "response"){

  V <- length(unique(view))
  n <- nrow(newx)
  metadat <- object$metadat
  Z <- matrix(NA, n, V)
  for (v in 1:V){
    Z[,v] <- as.matrix(cbind(1, newx[, view == v, drop=FALSE]) %*% object$base[[v]])
  }

  if(metadat == "response"){
    Z <- 1/(1+exp(-Z))
  }
  else if(metadat == "class"){
    Z <- 1*(1/(1+exp(-Z)) > 0.5)
  }
  else if(metadat != "link"){
    stop("metadat should be one of 'response', 'class' or 'link'.")
  }

  if(!is.null(newcf)){
    Z <- cbind(newcf, Z)
  }

  out <- as.matrix(cbind(1, Z) %*% object$meta)

  if(predtype == "response"){
    out <- 1/(1+exp(-out))
  }
  else if(predtype == "class"){
    out <- 1*(1/(1+exp(-out)) > 0.5)
  }

  else if(predtype != "link"){
    stop("predtype should be one of 'response', 'class' or 'link'.")
  }

  return(out)

}
